### PR TITLE
refactor: Remove `MutableBufferDb` and Partition

### PR DESCRIPTION
**Pull request by alamb**
# Rationale

In #1057 we moved the responsibility for storage of mutable buffer Chunk data to the Catalog. In this PR we remove the last vestiges of Databases and partitions from the mutable buffer into DB.

The sequence for  https://github.com/influxdata/influxdb_iox/issues/1042 looks like

- [x] Storage directly in the catalog (https://github.com/influxdata/influxdb_iox/pull/1057)
- [x] Remove  `MutableBufferDB` (this PR)
- [ ] Switch to using ReadBuffer chunks directly in the catalog (not read buffer DB)
- [ ] Remove ReadBufferDB

### CODE REVIEW

File: server/src/db.rs [hunk 1]
```diff
@@ -394,6 +437,35 @@ impl Db {
         self.chunks(partition_key).into_iter().map(|c| c.summary())
     }
 
+    /// Returns the partition_keys in the requested sort order
+    pub fn partition_keys_sorted_by(&self, sort_rules: &SortOrder) -> Vec<String> {
```
Comment (alamb) [inline server/src/db.rs:441]: this was ported from MutableBuffer DB

**Comment by tustvold** [comment]
Will take a more thorough look tomorrow if someone doesn't get there first, but for chunk movement I will need the sort_by and last write to be at the chunk level - I'm aware you're just moving code here, and so I'm happy to do it as part of my PR but thought I'd mention it

File: mutable_buffer/src/partition.rs [hunk 2]
```diff
@@ -1,983 +0,0 @@
...
-                    Some(&partition.open_chunk)
-                } else {
-                    None
-                }
-            })
-    }
-}
-
-#[cfg(test)]
-mod tests {
```
Comment (tustvold) [inline mutable_buffer/src/partition.rs:302]: I'm guessing these tests, or equivalents have already been ported over?

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Just a question of if we are losing some tests that could be ported across for some free additional catalog test coverage - otherwise looks great!

File: server/src/db.rs [hunk 3]
```diff
@@ -495,10 +567,12 @@ impl Database for Db {
                                 .set_open(Chunk::new(chunk_id))
                                 .context(CreatingChunk { partition_key })?;
                         }
+                        partition.update_last_write_at();
                         chunk
                     }
                     Some(partition) => {
-                        let partition = partition.read();
+                        let mut partition = partition.write();
+                        partition.update_last_write_at();
```
Comment (tustvold) [inline server/src/db.rs:575]: There is a TODO a couple of lines down that could be removed

File: mutable_buffer/src/partition.rs (hunk 2 above)
Comment (alamb) [reply mutable_buffer/src/partition.rs:302]: yes, sorry -- I reviewed these tests and ported the ones that didn't have equivalents. I should have annotated them in the initial PR and I will do so now

File: mutable_buffer/src/partition.rs [hunk 4]
```diff
@@ -1,983 +0,0 @@
...
-    use data_types::partition_metadata::PartitionSummary;
-    use internal_types::{data::split_lines_into_write_entry_partitions, selection::Selection};
-
-    use arrow_deps::{
-        arrow::record_batch::RecordBatch, assert_table_eq, test_util::sort_record_batch,
-    };
-    use influxdb_line_protocol::parse_lines;
-
-    #[tokio::test]
-    async fn test_rollover_chunk() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:314]: rollover is well tested in both the `Db` tests as well as `query_tests`

File: mutable_buffer/src/partition.rs [hunk 5]
```diff
@@ -1,983 +0,0 @@
...
-                                          // can't drop again
-        let e = partition.drop_chunk(0).unwrap_err();
-        assert_eq!(
-            "Can not drop unknown chunk '0' of partition with key 'a_key' in mutable buffer. Valid chunk ids: [1]",
-            format!("{}", e)
-        );
-    }
-
-    #[tokio::test]
-    async fn test_chunk_timestamps() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:583]: this test was ported over to db.rs

File: mutable_buffer/src/partition.rs [hunk 6]
```diff
@@ -1,983 +0,0 @@
...
-        let after_data_load_2 = Utc::now();
-        let chunk = partition.rollover_chunk();
-
-        assert!(chunk.time_of_first_write.unwrap() < after_data_load_1);
-        assert!(chunk.time_of_first_write.unwrap() < chunk.time_of_last_write.unwrap());
-        assert!(chunk.time_of_last_write.unwrap() < after_data_load_2);
-    }
-
-    #[tokio::test]
-    async fn test_chunk_timestamps_empty() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:635]: This test no longer makes sense as you can't create empty Chunks anymore (they are only created when data is written)

File: mutable_buffer/src/partition.rs [hunk 7]
```diff
@@ -1,983 +0,0 @@
...
-        let chunk = partition.rollover_chunk();
-        let after_rollover = Utc::now();
-        assert!(chunk.time_of_first_write.is_none());
-        assert!(chunk.time_of_last_write.is_none());
-        assert!(after_partition_creation < chunk.time_closed.unwrap());
-        assert!(chunk.time_closed.unwrap() < after_rollover);
-    }
-
-    #[tokio::test]
-    async fn test_chunk_timestamps_empty_write() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:648]: I felt the additional coverage of these tests for timestamps was not worth porting over. While the chunk can handle writes without any data, that is not a case where the behavior is important and thus testing to ensure it doesn't change seems unecessary

File: mutable_buffer/src/partition.rs [hunk 8]
```diff
@@ -1,983 +0,0 @@
...
-        let chunk0_rollover = partition.rollover_chunk();
-        // old data remains unchanged
-        assert_table_eq!(expected0, &dump_chunk_table(&chunk0_snapshot0, "h2o"));
-        assert_table_eq!(expected1, &dump_chunk_table(&chunk0_snapshot1, "h2o"));
-        assert_table_eq!(expected2, &dump_chunk_table(&chunk0_snapshot2, "h2o"));
-        assert_table_eq!(expected2, &dump_chunk_table(&chunk0_rollover, "h2o"));
-    }
-
-    #[tokio::test]
-    async fn partition_size() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:836]: there is no `size` method in the catalog partition, instead there is a size field on `ChunkSummary` which is already tested

File: mutable_buffer/src/partition.rs [hunk 9]
```diff
@@ -1,983 +0,0 @@
...
-        // increase by the same initial amount
-        let chunk = partition.rollover_chunk();
-        assert_eq!(323, chunk.size());
-
-        load_data(&mut partition, &["h2o,state=MA,city=Boston temp=71.4 100"]).await;
-        assert_eq!(459, partition.size());
-    }
-
-    #[tokio::test]
-    async fn chunk_summaries_return_open_and_closed() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:868]: chunk summaries are tested in db.rs

File: mutable_buffer/src/partition.rs [hunk 10]
```diff
@@ -1,983 +0,0 @@
...
-        assert_eq!(3, summaries.len());
-
-        let summary = PartitionSummary::from_table_summaries(&partition.key, summaries);
-        assert_eq!(2, summary.tables.len());
-        assert!(summary.table("cpu").is_some());
-        assert!(summary.table("mem").is_some());
-    }
-
-    #[tokio::test]
-    async fn write_updates_last_write_at() {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:896]: this was ported

**Comment by alamb** [comment]
Sending this in so that @tustvold  can build automated chunk movement off of it

**Pull request merged**